sfge/src/main/java/com/salesforce/rules/UnusedMethodRule.java

```diff
@@ -1,155 +1,159 @@
 import com.salesforce.graph.ops.directive.EngineDirective;
-import com.salesforce.graph.vertex.MethodVertex;
-import com.salesforce.graph.vertex.SFVertexFactory;
-import com.salesforce.rules.unusedmethod.*;
+import com.salesforce.graph.vertex.*;
 import java.util.*;
-import java.util.stream.Collectors;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
-import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversal;
 import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversalSource;
-import org.apache.tinkerpop.gremlin.structure.Vertex;
 
-public class UnusedMethodRule extends AbstractStaticRule {
+/**
+ * Rule for identifying dead code. After all {@link ApexPath} instances have been traversed,
+ * generates violations for any eligible {@link MethodVertex} that was never invoked. A method is
+ * considered eligible if it meets all the following criteria:
+ *
+ * <ol>
+ *   <li>The method is not declared as a {@code testMethod}
+ *   <li>The method is not declared as {@code abstract}
+ *   <li>The method is not a private, 0-arity constructor
+ *   <li>The method is not a VisualForce getter/setter
+ *   <li>The method is not considered a path entrypoint
+ *   <li>The method is not annotated with an engine directive that silences this rule.
+ * </ol>
+ */
+public final class UnusedMethodRule extends AbstractPathTraversalRule
+        implements PostProcessingRule {
+    private static final Logger LOGGER = LogManager.getLogger(UnusedMethodRule.class);
     private static final String URL =
             "https://forcedotcom.github.io/sfdx-scanner/en/v3.x/salesforce-graph-engine/rules/#UnusedMethodRule";
-    private static final Logger LOGGER = LogManager.getLogger(UnusedMethodRule.class);
-    private static final String DESCRIPTION = "Identifies methods that are not invoked";
-    private static final String VIOLATION_TEMPLATE = "Method %s in class %s is never invoked";
-
-    GraphTraversalSource g;
-    /** A helper object used to track state and caching as the rule executes. */
-    RuleStateTracker ruleStateTracker;
+    /**
+     * Used internally for tracking every unique method invoked during the traversal of every path.
+     */
+    private Set<String> encounteredUsageKeys;
 
     private UnusedMethodRule() {
-        super();
-    }
-
-    public static UnusedMethodRule getInstance() {
-        return LazyHolder.INSTANCE;
-    }
-
-    @Override
-    protected int getSeverity() {
-        return SEVERITY.LOW.code;
-    }
-
-    @Override
-    protected String getDescription() {
-        return DESCRIPTION;
+        this.encounteredUsageKeys = CollectionUtil.newTreeSet();
     }
 
     @Override
-    protected String getCategory() {
-        return CATEGORY.PERFORMANCE.name;
+    public boolean test(BaseSFVertex vertex) {
+        // This rule is interested in any vertex representing an invocation of a method.
+        return vertex instanceof InvocableVertex;
```

Comment:
If an invocable is in the path, this should always have returned true, right? I'm still trying to wrap my head around how an invocable is not called here and we require an additional subpath search. Is there another category that's missing here and all we need to do is to identify it and return true?

Reply:
In any case, this would probably still not include skipped method call paths that had anomalies such as NPE and recursion.